library/alloc/src/str.rs

```diff
@@ -619,3 +637,60 @@ impl str {
 pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
     unsafe { Box::from_raw(Box::into_raw(v) as *mut str) }
 }
+
+/// Converts the bytes while the bytes are still ascii.
+/// This is performed in chunks of `16*size_of::<usize>()` which in testing
+/// performed the fastest.
+#[inline]
+#[cfg(not(test))]
+#[cfg(not(no_global_oom_handling))]
+fn convert_while_ascii(b: &[u8], out: &mut Vec<u8>, convert: fn(&u8) -> u8) {
+    out.reserve(b.len());
+
+    const USIZE_SIZE: usize = mem::size_of::<usize>();
+    const MAGIC_UNROLL: usize = 16;
+
+    let mut i = 0;
+    while i + USIZE_SIZE * MAGIC_UNROLL <= b.len() {
+        unsafe {
+            let range = i..i + USIZE_SIZE * MAGIC_UNROLL;
+
+            // Safety: we have checks the sizes `b` and `out` to know that our
+            let in_chunk = b.get_unchecked(range.clone());
+            let out_chunk = out.spare_capacity_mut().get_unchecked_mut(range);
+
+            let mut count = 0;
+            for j in 0..MAGIC_UNROLL {
+                // read the bytes 1 usize at a time (unaligned since we haven't checked the alignment)
+                // safety: in_chunk is valid bytes in the range
+                let chunk = in_chunk.as_ptr().cast::<usize>().add(j).read_unaligned();
+                count += contains_nonascii(chunk) as usize;
+            }
+            if count > 0 {
+                // safety: we know that i bytes are initialised
+                out.set_len(i);
+                return;
+            }
+
+            // perform the case conversions on USIZE_SIZE * MAGIC_UNROLL bytes (gets heavily autovec'd)
+            for j in 0..USIZE_SIZE * MAGIC_UNROLL {
+                // safety: in_chunk and out_chunk is valid bytes in the range
+                let out = out_chunk.get_unchecked_mut(j);
+                out.write(convert(in_chunk.get_unchecked(j)));
+            }
+        }
+
+        // mark these bytes as initialised
+        i += USIZE_SIZE * MAGIC_UNROLL;
+    }
+
+    // safety: we know that i bytes are initialised
+    unsafe { out.set_len(i) };
+
+    /// Returns `true` if any byte in the word `v` is nonascii (>= 128). Stolen from core::slice::is_ascii
+    #[inline]
+    fn contains_nonascii(v: usize) -> bool {
+        const NONASCII_MASK: usize = 0x8080808080808080_u64 as usize;
+        (NONASCII_MASK & v) != 0
+    }
+}
```

Comment:
Problem is that you hardcoded `usize` size for 64bit.

Reply:
This should fix that:
```rust
const NONASCII_MASK: usize = usize::from_ne_bytes([0x80; core::mem::size_of::<usize>()]);
```

Reply:
>  Or perhaps not, this will be trimmed on smaller usizes.

Yeah, it does get trimmed. But I'll do the `from_ne_bytes` method just to be safe

Reply:
https://github.com/rust-lang/rust/blob/653463731a7f01f519cf85f444869def27f00395/library/core/src/num/mod.rs#L895-L898

Reply:
That fn is from code, and is not accessible in alloc